### ISSUE TITLE
Poor residual learning reward when using the new DexHand DG-5F

### ISSUE THREAD

**Issue by AaronChenNMNL** [ManipTrans/ManipTrans#47]
Thank you for your awesome work!

I’ve been experimenting with applying ManipTrans on the new dexterous hand, DG-5F.
The `data_index` I used is `g0`.

The imitator module trains well and achieves good performance (reward > 60) during the Hand Trajectory Imitation stage.
However, the residual module performs significantly worse — the reward starts around 6 and never exceeds 20 even after training.
All inference rollouts also failed to perform the task.

Below is one of the rollout examples.

https://github.com/user-attachments/assets/94a1ec5b-e0d4-46a4-80ed-8e4cbf42742d

I’ve also attached the relevant files:

- `dg5f.py`
- maniptrans_envs/assets/dg5f_hand` (taken from the [repo](https://github.com/Tesollo-Delto/DELTO_M_ROS2), but with the up_axis in the DAE file changed from Z_UP to Y_UP)

My assumption is that the DG-5F might experience a similar issue as the Allegro hand, which can be mitigated by tuning the `relative_translation` parameter.
Since I noticed that `grab_dataset_dexhand.py` does not take this parameter into account, I tried training with the `data_idx` `e9aab@2`, where this parameter is applied.
Nevertheless, during the **Hand Trajectory Imitation stage**, the reward remains stuck around 24, which I suspect may indicate that the current `relative_translation` value is not suitable for DG-5F.

I have two questions:

1. Do you have any suggestions on what might be causing the poor performance of the residual module? Could it be related to the `relative_translation` parameter, or might there be something incorrect in my `dg5f.py` setup?
2. How should I properly tune the `relative_translation` parameter? I’ve read the related [issue](https://github.com/ManipTrans/ManipTrans/issues/25#issuecomment-2858336232) but found it quite difficult to tune in practice since the hand keeps moving.

Thanks again for your time and for maintaining this great project!

Best,
Aaron

[dg5f.py](https://github.com/user-attachments/files/22801182/dg5f.py)

[dg5f_hand.zip](https://github.com/user-attachments/files/22801195/dg5f_hand.zip)

**Comment by KailinLi** [comment]
Hi Aaron,

Thanks for your interest in ManipTrans and for extending the framework to the DG-5F. Because the DG-5F differs substantially from a human hand in morphology, we recommend slightly relaxing the tracking-failure criterion (see the code around [this line](https://github.com/ManipTrans/ManipTrans/blob/a3d08cfe3c3a5868a7f057533bcaf759c5af4705/maniptrans_envs/lib/envs/tasks/dexhandmanip_sh.py#L1531)).

https://github.com/user-attachments/assets/d77f9fff-faa0-4fad-8d2c-012523f733fb

The video above is my quick result on `g0`; I didn’t spend much time on hand-specific tuning. With more parameter adjustments, the performance should improve further.

If anything is unclear, feel free to follow up with details—we’re happy to help.

**Comment by AaronChenNMNL** [comment]
Thank you so much for your quick reply and helpful suggestions!

I’ll try adjusting the `scale_factor` to relax the tracking-failure criterion.

I’ve also encountered a performance issue when training the **imitator** for DG-5F on the data `20aed@0` — the reward only reaches around 25.46 after roughly 4 hours of training.
In contrast, the Allegro and Inspire hands can easily achieve rewards above 100 within just 30 minutes.

Even after tuning the `relative_translation` for DG-5F (I found that values between [0.0, 0.0, -0.085] and [0.0, 0.0, -0.095] seem most suitable), the reward still doesn’t exceed 25 after one hour of training.

Could this performance gap also be related to the morphological differences between the human hand and the DG-5F?
I’ll continue experimenting with the tracking-failure settings in `dexhand_imitator.py` to see if this helps improve the imitator’s performance.
Any additional suggestions or insights would be greatly appreciated!

Thanks again for your help!

**Comment by KailinLi** [comment]
Hi Aaron,

Yes—I expect relaxing the tracking-failure condition (e.g., via scale_factor) to make training smoother on DG-5F.

A quick clarification: as shown in our paper, the imitator is intended to be trained on a large set of manipulation sequences. Training on a single sequence such as `20aed@0` won’t reflect the intended design and can plateau at low reward. If you just need a fast single-sequence sanity check, you can skip the imitator and train only the residual model by setting `rh_base_model_checkpoint = null` during residual training.

For the low reward on `20aed@0`, I can’t diagnose precisely, but here’s a short checklist—please enable visualization (`headless = false`) and verify:
	1.	Any DG-5F finger self-intersections?
	2.	Severe hand–object interpenetration (considering DG-5F is relatively large)?
	3.	Finger–table intersections during reaching?
	4.	Whether mano2dexhand.py ran correctly (axis, joint mapping, limits).
	5.	If fingertips don’t actually touch the object, gently increase the contact-reward weight/threshold.
	6.	For bimanual tasks, whether the two hands collide with each other.

In general, the larger the morphology gap from a human hand, the harder the transfer—so a bit more tuning is usually needed.

**Comment by AaronChenNMNL** [comment]
Thank you for your insightful suggestions!

I’m currently trying two things out.

First, I’m attempting to reproduce the successful rollout shown in your _dg5f.mp4._ I adjusted the `scale_factor` in `dexhandmanip_sh.py` to 2.0 for both training and testing with `data_idx = g0`. Below are two approaches I’ve tried:

1. Training the imitator first, then the residual module. The imitator and residual modules reach rewards of approximately 60 and 45, respectively. However, the rollout results look quite different from yours.

https://github.com/user-attachments/assets/085f842e-c839-4e73-9d66-d040eb877a08

2. Training the residual module from scratch. This setup hasn’t achieved a reward higher than 45 after 4 hours of training.
```
python main/dataset/mano2dexhand.py --data_idx g0 --dexhand dg5f --headless --iter 4000
python main/rl/train.py task=ResDexHand dexhand=dg5f side=RH headless=true num_envs=8192 learning_rate=2e-4 test=false randomStateInit=true rh_base_model_checkpoint=null lh_base_model_checkpoint=null dataIndices=[g0] early_stop_epochs=2000 actionsMovingAverage=0.4 experiment=cross_g0_dg5f sim_device=cuda:0 rl_device=cuda:0
```

Would you mind sharing a few details about how you obtained the successful rollout?

Second, I followed the checklist you provided on `20aed@0`:

1. There are some self-intersections, mainly between the middle and ring fingers.
2. Severe hand–object interpenetration sometimes occurs, where the lid penetrates into the DG-5F.
3. I didn’t observe any finger–table intersections for `20aed@0`.
4. `mano2dexhand.py` seems to work correctly.
5. The fingertips do make contact with the object.

Below are the results from running `mano2dexhand.py` for 4000 iterations, along with a short video showing the training process for DG-5F and Allegro, respectively.

<img width="480" height="530" alt="Image" src="https://github.com/user-attachments/assets/b78e8978-6a15-46dc-8190-1daea5ab77e1" />

<img width="530" height="513" alt="Image" src="https://github.com/user-attachments/assets/3e1ef47e-dfe3-4882-a4b2-b9ce62eb6711" />

https://github.com/user-attachments/assets/fd75154a-fd0e-4a44-a67d-8c22499ffd8f

https://github.com/user-attachments/assets/3b53ac91-d01f-463f-9d67-97cc7b25ca66

I’ll continue experimenting with visualization to see whether there are any critical differences between different hands during preprocessing and training.

Thanks again for your time and detailed guidance!

**Comment by KailinLi** [comment]
Hi Aaron,

Thanks for the thorough follow-up!

For the `g0` case, I made three small edits to get the rollout you saw:

1. Relax **only** hand-keypoint failure terms (keep object terms unchanged). This keeps object motion closer to the reference. Concretely, my `failed_execute` looked like:
```
    failed_execute = (
        (
            (diff_obj_pos_dist > 0.02 / 0.343 * scale_factor**3)
            | (diff_thumb_tip_pos_dist > 0.04 * 2 / 0.7 * scale_factor)
            | (diff_index_tip_pos_dist > 0.045 * 2 / 0.7 * scale_factor)
            | (diff_middle_tip_pos_dist > 0.05 * 2 / 0.7 * scale_factor)
            | (diff_pinky_tip_pos_dist > 0.06 * 2 / 0.7 * scale_factor)
            | (diff_ring_tip_pos_dist > 0.06 * 2 / 0.7 * scale_factor)
            | (diff_level_1_pos_dist > 0.07 * 2 / 0.7 * scale_factor)
            | (diff_level_2_pos_dist > 0.08 * 2 / 0.7 * scale_factor)
            | (diff_obj_rot_angle.abs() / np.pi * 180 > 30 / 0.343 * scale_factor**3)
            | torch.any((finger_tip_distance < 0.005) & ~(target_states["tip_contact_state"].any(1)), dim=-1)
        )
        & (running_progress_buf >= 8)
    ) | error_buf
```
2. Comment out a leftover trick in `mano2dexhand.py`. Specifically, comment the line around [here](https://github.com/ManipTrans/ManipTrans/blob/a3d08cfe3c3a5868a7f057533bcaf759c5af4705/main/dataset/mano2dexhand.py#L159) (it helped Inspire Hand but hurts DG-5F).
3. In `dg5f.py`, explicitly set `self.self_collision = False` to prevent finger–finger interpenetration.

With these three changes, you should be able to reproduce my dg5f.mp4 result.

For `20aed@0`, the same fixes should apply. In addition, consider gently increasing the contact-related reward to make sure fingertips are truly making/maintaining contact.

**Comment by AaronChenNMNL** [comment]
I apologize for the late reply, and thank you so much for your detailed help!
I’ve successfully reproduced the rollout results for data `g0`.
I was planning to reply after making some progress on improving the performance for data `20aed@0`, but it turned out to be not quite straightforward. I suspect the issue might lie in the preprocessing step, which could make the initial state for residual learning difficult to learn the task properly. I’m currently experimenting with the weights in `mano2dexhand.py` to see if that helps.

I’ll close this issue for now and reopen it if I encounter any further problems or follow-ups.
Any suggestions or insights would still be greatly appreciated.
Thanks again for your help and for such a wonderful project!
